chia/wallet/wallet_state_manager.py

```diff
@@ -641,68 +644,85 @@ async def unconfirmed_removals_for_wallet(self, wallet_id: int) -> Dict[bytes32,
 
     async def determine_coin_type(
         self, peer: WSChiaConnection, coin_state: CoinState, fork_height: Optional[uint32]
-    ) -> Optional[WalletIdentifier]:
+    ) -> Tuple[Optional[WalletIdentifier], Optional[VersionedBlob]]:
         if coin_state.created_height is not None and (
             self.is_pool_reward(uint32(coin_state.created_height), coin_state.coin)
             or self.is_farmer_reward(uint32(coin_state.created_height), coin_state.coin)
         ):
-            return None
+            return None, None
 
         response: List[CoinState] = await self.wallet_node.get_coin_state(
             [coin_state.coin.parent_coin_info], peer=peer, fork_height=fork_height
         )
         if len(response) == 0:
             self.log.warning(f"Could not find a parent coin with ID: {coin_state.coin.parent_coin_info}")
-            return None
+            return None, None
         parent_coin_state = response[0]
         assert parent_coin_state.spent_height == coin_state.created_height
 
         coin_spend = await fetch_coin_spend_for_coin_state(parent_coin_state, peer)
         if coin_spend is None:
-            return None
-
-        puzzle = Program.from_bytes(bytes(coin_spend.puzzle_reveal))
+            return None, None
 
+        puzzle = coin_spend.puzzle_reveal.to_program()
         uncurried = uncurry_puzzle(puzzle)
 
         # Check if the coin is a CAT
         cat_curried_args = match_cat_puzzle(uncurried)
         if cat_curried_args is not None:
+            cat_mod_hash, tail_program_hash, cat_inner_puzzle = cat_curried_args
+            cat_data: CATCoinData = CATCoinData(
+                bytes32(bytes(cat_mod_hash)[1:]), bytes32(bytes(tail_program_hash)[1:]), cat_inner_puzzle
```

Comment:
Similar to the comment above, you don't want to slice the bytes manually here.  Instead, do `<program>.atom`.